File: src/toubkal/app/core/diagnostics/logger.ts
/**
 * Toubkal Browser - Diagnostics Logger
 * 
 * Structured logging system for TypeScript/React components with privacy-safe logging
 * and integration with C++ logging system via Mojo IPC.
 */

// Browser APIs

export enum LogLevel {
  DEBUG = 0,
  INFO = 1,
  WARN = 2,
  ERROR = 3,
  FATAL = 4
}

export interface LogContext {
  [key: string]: unknown;
}

export interface LogEntry {
  level: LogLevel;
  component: string;
  message: string;
  context?: LogContext;
  correlationId?: string;
  timestamp: string;
}

export interface LoggerConfig {
  consoleEnabled?: boolean;
  fileEnabled?: boolean;
  jsonEnabled?: boolean;
  filePath?: string;
  jsonPath?: string;
  maxLogLevel?: LogLevel;
  privacyMode?: boolean;
}

export class Logger {
  private static instance: Logger;
  private config: LoggerConfig;
  private correlationId: string = '';
  private logBuffer: LogEntry[] = [];
  private maxBufferSize: number = 1000;

  private constructor() {
    this.config = {
      consoleEnabled: true,
      fileEnabled: false,
      jsonEnabled: false,
      maxLogLevel: LogLevel.INFO,
      privacyMode: true
    };
  }

  public static getInstance(): Logger {
    if (Logger.instance == null) {
      Logger.instance = new Logger();
    }
    return Logger.instance;
  }

  public initialize(config: Partial<LoggerConfig>): void {
    this.config = { ...this.config, ...config };
    
    // Setup global error handlers if enabled
    if (this.config.consoleEnabled === true) {
      this.setupGlobalErrorHandlers();
    }
  }

  public log(level: LogLevel, component: string, message: string, context: LogContext = {}): void {
    // Check log level filter - only log if level >= maxLogLevel
    const maxLevel = this.config.maxLogLevel ?? LogLevel.INFO;
    console.log(`Logger.log: level=${level}, maxLevel=${maxLevel}, shouldLog=${level >= maxLevel}`);
    if (level < maxLevel) {
      console.log('Logger.log: filtered out due to level');
      return;
    }

    const entry: LogEntry = {
      level,
      component,
      message,
      context: (this.config.privacyMode ?? false) === true ? this.redactPII(context) : context,
      correlationId: this.correlationId,
      timestamp: new Date().toISOString()
    };

    // Add to buffer
    this.logBuffer.push(entry);
    if (this.logBuffer.length > this.maxBufferSize) {
      this.logBuffer.shift(); // Remove oldest entry
    }

    // Output to configured sinks
    if (this.config.consoleEnabled === true) {
      this.logToConsole(entry);
    }

    if (this.config.fileEnabled === true && (this.config.filePath ?? '').length > 0) {
      void this.logToFile(entry);
    }

    if (this.config.jsonEnabled === true && (this.config.jsonPath ?? '').length > 0) {
      this.logToJson(entry);
    }
  }

  public debug(component: string, message: string, context: LogContext = {}): void {
    this.log(LogLevel.DEBUG, component, message, context);
  }

  public info(component: string, message: string, context: LogContext = {}): void {
    this.log(LogLevel.INFO, component, message, context);
  }

  public warn(component: string, message: string, context: LogContext = {}): void {
    this.log(LogLevel.WARN, component, message, context);
  }

  public error(component: string, message: string, context: LogContext = {}): void {
    this.log(LogLevel.ERROR, component, message, context);
  }

  public fatal(component: string, message: string, context: LogContext = {}): void {
    this.log(LogLevel.FATAL, component, message, context);
  }

  public setCorrelationId(correlationId: string): void {
    this.correlationId = correlationId;
  }

  public getCorrelationId(): string {
    return this.correlationId;
  }

  public getLogBuffer(): LogEntry[] {
    return [...this.logBuffer];
  }

  public clearLogBuffer(): void {
    this.logBuffer = [];
  }

  public exportLogs(): string {
    return JSON.stringify(this.logBuffer, null, 2);
  }

  public getRecentLogs(limit: number = 100): LogEntry[] {
    return this.logBuffer.slice(-limit);
  }

  public clearLogs(): void {
    this.logBuffer = [];
  }

  public getConfig(): LoggerConfig {
    return { ...this.config };
  }

  public isConsoleEnabled(): boolean {
    return (this.config.consoleEnabled ?? false) === true;
  }

  public isFileEnabled(): boolean {
    return (this.config.fileEnabled ?? false) === true;
  }

  public isJsonEnabled(): boolean {
    return (this.config.jsonEnabled ?? false) === true;
  }

  public getMaxLogLevel(): LogLevel {
    return this.config.maxLogLevel ?? LogLevel.INFO;
  }

  public isPrivacyMode(): boolean {
    return (this.config.privacyMode ?? false) === true;
  }

  public getHealthStatus(): { status: string; message: string; details: Record<string, unknown> } {
    return {
      status: 'healthy',
      message: 'Logger is operational',
      details: {
        bufferSize: this.logBuffer.length,
        maxBufferSize: this.maxBufferSize,
        consoleEnabled: this.config.consoleEnabled,
        fileEnabled: this.config.fileEnabled,
        jsonEnabled: this.config.jsonEnabled,
        privacyMode: this.config.privacyMode
      }
    };
  }

  private logToConsole(entry: LogEntry): void {
    const levelStr = this.getLevelString(entry.level);
    const timestamp = entry.timestamp;
    const contextStr = (entry.context != null && Object.keys(entry.context).length > 0) 
      ? ` ${JSON.stringify(entry.context)}` 
      : '';
    const correlationStr = (entry.correlationId ?? '').length > 0 ? ` [${entry.correlationId}]` : '';

    const logMessage = `[${timestamp}] ${levelStr} ${entry.component}: ${entry.message}${contextStr}${correlationStr}`;

    // Use appropriate console method based on level
    switch (entry.level) {
      case LogLevel.DEBUG:
        console.debug(logMessage);
        break;
      case LogLevel.INFO:
        console.info(logMessage);
        break;
      case LogLevel.WARN:
        console.warn(logMessage);
        break;
      case LogLevel.ERROR:
        console.error(logMessage);
        break;
      case LogLevel.FATAL:
        console.error(`FATAL: ${logMessage}`);
        break;
    }
  }

  private logToFile(entry: LogEntry): void {
    if ((this.config.filePath ?? '').length === 0) return;

    try {
      // Use File System Access API if available (with user consent)
      if ('showSaveFilePicker' in window) {
        // This would require user interaction for file access
        // For now, we'll use a different approach
        this.logToLocalStorage(entry);
      } else {
        this.logToLocalStorage(entry);
      }
    } catch (error) {
      console.error('[Logger] Failed to write to file:', error);
    }
  }

  private logToJson(entry: LogEntry): void {
    if ((this.config.jsonPath ?? '').length === 0) return;

    try {
      const jsonStr = JSON.stringify(entry) + '\n';
      // Store in localStorage for now (would be replaced with proper file access)
      const existing = localStorage.getItem('toubkal_logs_json') ?? '';
      localStorage.setItem('toubkal_logs_json', existing + jsonStr);
    } catch (error) {
      console.error('[Logger] Failed to write JSON log:', error);
    }
  }

  private logToLocalStorage(entry: LogEntry): void {
    try {
      const existing = localStorage.getItem('toubkal_logs') ?? '';
      const logLine = `[${entry.timestamp}] ${this.getLevelString(entry.level)} ${entry.component}: ${entry.message}\n`;
      localStorage.setItem('toubkal_logs', existing + logLine);
    } catch (error) {
      console.error('[Logger] Failed to write to localStorage:', error);
    }
  }

  private setupGlobalErrorHandlers(): void {
    // Global error handler
    window.addEventListener('error', (event) => {
      this.error('GlobalError', event.message, {
        filename: event.filename,
        lineno: event.lineno,
        colno: event.colno,
        stack: event.error?.stack
      });
    });

    // Unhandled promise rejection handler
    window.addEventListener('unhandledrejection', (event) => {
      this.error('UnhandledRejection', 'Unhandled promise rejection', {
        reason: event.reason,
        stack: event.reason?.stack
      });
    });
  }

  private getLevelString(level: LogLevel): string {
    switch (level) {
      case LogLevel.DEBUG: return 'DEBUG';
      case LogLevel.INFO: return 'INFO';
      case LogLevel.WARN: return 'WARN';
      case LogLevel.ERROR: return 'ERROR';
      case LogLevel.FATAL: return 'FATAL';
      default: return 'UNKNOWN';
    }
  }

  private redactPII(context: LogContext): LogContext {
    const redacted = { ...context };
    
    // List of PII fields to redact (excluding url which has special handling)
    const piiFields = [
      'email', 'phone', 'ssn', 'creditCard', 'password', 'token',
      'userId', 'sessionId', 'ipAddress', 'query', 'search',
      'name', 'address', 'birthday', 'socialSecurity'
    ];

    for (const field of piiFields) {
      if (field in redacted) {
        redacted[field] = '[REDACTED]';
      }
    }

    // Redact URLs and queries (special handling for url field)
    if (typeof redacted.url === 'string') {
      try {
        const url = new URL(redacted.url as string);
        // Only redact query parameters, keep the path
        redacted.url = `${url.protocol}//${url.hostname}${url.pathname}`;
      } catch {
        redacted.url = '[REDACTED_URL]';
      }
    }

    return redacted;
  }
}

// Convenience functions
export const logger = Logger.getInstance();

// Convenience functions for common logging patterns
export const logDebug = (component: string, message: string, context?: LogContext) => {
  logger.debug(component, message, context);
};

export const logInfo = (component: string, message: string, context?: LogContext) => {
  logger.info(component, message, context);
};

export const logWarn = (component: string, message: string, context?: LogContext) => {
  logger.warn(component, message, context);
};

export const logError = (component: string, message: string, context?: LogContext) => {
  logger.error(component, message, context);
};

export const logFatal = (component: string, message: string, context?: LogContext) => {
  logger.fatal(component, message, context);
};

// Privacy-safe logging functions
export const logSafeDebug = (component: string, message: string, context?: LogContext) => {
  logger.debug(component, message, context);
};

export const logSafeInfo = (component: string, message: string, context?: LogContext) => {
  logger.info(component, message, context);
};

export const logSafeWarn = (component: string, message: string, context?: LogContext) => {
  logger.warn(component, message, context);
};

export const logSafeError = (component: string, message: string, context?: LogContext) => {
  logger.error(component, message, context);
};

export const logSafeFatal = (component: string, message: string, context?: LogContext) => {
  logger.fatal(component, message, context);
};

// Export missing classes for tests
export { ErrorTracker } from './error-tracker';
export { PerformanceMonitor } from './performance-monitor';
export { ScalabilityManager } from './scalability-manager';

File: src/toubkal/components/diagnostics/logging/logger.cc
// Copyright 2025 Toubkal Browser. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "toubkal/components/diagnostics/logging/logger.h"

#include <fstream>
#include <iomanip>
#include <sstream>

#include "base/files/file_util.h"
#include "base/json/json_writer.h"
#include "base/logging.h"
#include "base/rand_util.h"
#include "base/strings/string_util.h"
#include "base/strings/stringprintf.h"
#include "base/time/time.h"

namespace toubkal {
namespace diagnostics {

// ConsoleLogSink implementation
void ConsoleLogSink::Write(const LogEntry& entry) {
  std::string level_str;
  switch (entry.level) {
    case LogLevel::DEBUG:
      level_str = "DEBUG";
      break;
    case LogLevel::INFO:
      level_str = "INFO";
      break;
    case LogLevel::WARN:
      level_str = "WARN";
      break;
    case LogLevel::ERROR:
      level_str = "ERROR";
      break;
    case LogLevel::FATAL:
      level_str = "FATAL";
      break;
  }

  std::string timestamp_str = base::TimeFormatWithPattern(
      entry.timestamp, "yyyy-MM-dd HH:mm:ss.SSS");
  
  std::string context_str;
  if (!entry.context.empty()) {
    std::string json_str;
    base::JSONWriter::Write(entry.context, &json_str);
    context_str = " " + json_str;
  }
  
  std::string correlation_str;
  if (!entry.correlation_id.empty()) {
    correlation_str = " [" + entry.correlation_id + "]";
  }

  std::string log_line = base::StringPrintf(
      "[%s] %s %s: %s%s%s",
      timestamp_str.c_str(),
      level_str.c_str(),
      entry.component.c_str(),
      entry.message.c_str(),
      context_str.c_str(),
      correlation_str.c_str());

  // Use Chromium's logging for console output
  switch (entry.level) {
    case LogLevel::DEBUG:
      VLOG(1) << log_line;
      break;
    case LogLevel::INFO:
      LOG(INFO) << log_line;
      break;
    case LogLevel::WARN:
      LOG(WARNING) << log_line;
      break;
    case LogLevel::ERROR:
      LOG(ERROR) << log_line;
      break;
    case LogLevel::FATAL:
      LOG(FATAL) << log_line;
      break;
  }
}

void ConsoleLogSink::Flush() {
  // Console output is automatically flushed
}

// FileLogSink implementation
FileLogSink::FileLogSink(const std::string& file_path) : file_path_(file_path) {
  DCHECK(!file_path_.empty());
}

FileLogSink::~FileLogSink() {
  Flush();
}

void FileLogSink::Write(const LogEntry& entry) {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  
  std::ofstream file(file_path_, std::ios::app);
  if (!file.is_open()) {
    LOG(ERROR) << "Failed to open log file: " << file_path_;
    return;
  }

  std::string level_str;
  switch (entry.level) {
    case LogLevel::DEBUG:
      level_str = "DEBUG";
      break;
    case LogLevel::INFO:
      level_str = "INFO";
      break;
    case LogLevel::WARN:
      level_str = "WARN";
      break;
    case LogLevel::ERROR:
      level_str = "ERROR";
      break;
    case LogLevel::FATAL:
      level_str = "FATAL";
      break;
  }

  std::string timestamp_str = base::TimeFormatWithPattern(
      entry.timestamp, "yyyy-MM-dd HH:mm:ss.SSS");
  
  std::string context_str;
  if (!entry.context.empty()) {
    std::string json_str;
    base::JSONWriter::Write(entry.context, &json_str);
    context_str = " " + json_str;
  }
  
  std::string correlation_str;
  if (!entry.correlation_id.empty()) {
    correlation_str = " [" + entry.correlation_id + "]";
  }

  file << base::StringPrintf(
      "[%s] %s %s: %s%s%s\n",
      timestamp_str.c_str(),
      level_str.c_str(),
      entry.component.c_str(),
      entry.message.c_str(),
      context_str.c_str(),
      correlation_str.c_str());
}

void FileLogSink::Flush() {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  // File is automatically flushed on close
}

// JsonLogSink implementation
JsonLogSink::JsonLogSink(const std::string& file_path) : file_path_(file_path) {
  DCHECK(!file_path_.empty());
}

JsonLogSink::~JsonLogSink() {
  Flush();
}

void JsonLogSink::Write(const LogEntry& entry) {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  
  std::ofstream file(file_path_, std::ios::app);
  if (!file.is_open()) {
    LOG(ERROR) << "Failed to open JSON log file: " << file_path_;
    return;
  }

  base::Value::Dict log_entry;
  log_entry.Set("timestamp", base::TimeFormatWithPattern(
      entry.timestamp, "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"));
  log_entry.Set("level", static_cast<int>(entry.level));
  log_entry.Set("component", entry.component);
  log_entry.Set("message", entry.message);
  
  if (!entry.context.empty()) {
    log_entry.Set("context", entry.context.Clone());
  }
  
  if (!entry.correlation_id.empty()) {
    log_entry.Set("correlation_id", entry.correlation_id);
  }

  std::string json_str;
  base::JSONWriter::Write(log_entry, &json_str);
  file << json_str << "\n";
}

void JsonLogSink::Flush() {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  // File is automatically flushed on close
}

// Logger implementation
Logger& Logger::GetInstance() {
  static Logger instance;
  return instance;
}

void Logger::Initialize(const base::Value::Dict& config) {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  
  // Clear existing sinks
  sinks_.clear();
  
  // Add console sink if enabled
  if (config.FindBool("console_enabled").value_or(true)) {
    AddSink(base::MakeRefCounted<ConsoleLogSink>());
  }
  
  // Add file sink if configured
  if (const std::string* file_path = config.FindString("file_path")) {
    AddSink(base::MakeRefCounted<FileLogSink>(*file_path));
  }
  
  // Add JSON sink if configured
  if (const std::string* json_path = config.FindString("json_path")) {
    AddSink(base::MakeRefCounted<JsonLogSink>(*json_path));
  }
}

void Logger::Log(LogLevel level, const std::string& component,
                 const std::string& message, const base::Value::Dict& context) {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  
  LogEntry entry(level, component, message);
  entry.context = context.Clone();
  entry.correlation_id = correlation_id_;
  
  for (const auto& sink : sinks_) {
    sink->Write(entry);
  }
}

void Logger::Debug(const std::string& component, const std::string& message,
                   const base::Value::Dict& context) {
  Log(LogLevel::DEBUG, component, message, context);
}

void Logger::Info(const std::string& component, const std::string& message,
                  const base::Value::Dict& context) {
  Log(LogLevel::INFO, component, message, context);
}

void Logger::Warn(const std::string& component, const std::string& message,
                  const base::Value::Dict& context) {
  Log(LogLevel::WARN, component, message, context);
}

void Logger::Error(const std::string& component, const std::string& message,
                   const base::Value::Dict& context) {
  Log(LogLevel::ERROR, component, message, context);
}

void Logger::Fatal(const std::string& component, const std::string& message,
                   const base::Value::Dict& context) {
  Log(LogLevel::FATAL, component, message, context);
}

void Logger::AddSink(scoped_refptr<LogSink> sink) {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  DCHECK(sink);
  sinks_.push_back(sink);
}

void Logger::RemoveSink(scoped_refptr<LogSink> sink) {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  auto it = std::find(sinks_.begin(), sinks_.end(), sink);
  if (it != sinks_.end()) {
    sinks_.erase(it);
  }
}

void Logger::LogSafe(LogLevel level, const std::string& component,
                     const std::string& message, const base::Value::Dict& context) {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  
  base::Value::Dict safe_context = RedactPII(context);
  Log(level, component, message, safe_context);
}

void Logger::SetCorrelationId(const std::string& correlation_id) {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  correlation_id_ = correlation_id;
}

std::string Logger::GetCorrelationId() const {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  return correlation_id_;
}

void Logger::Flush() {
  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
  for (const auto& sink : sinks_) {
    sink->Flush();
  }
}

base::Value::Dict Logger::RedactPII(const base::Value::Dict& context) {
  base::Value::Dict redacted = context.Clone();
  
  // Redact common PII fields
  const std::vector<std::string> pii_fields = {
    "email", "phone", "ssn", "credit_card", "password", "token",
    "user_id", "session_id", "ip_address", "url", "query"
  };
  
  for (const auto& field : pii_fields) {
    if (redacted.contains(field)) {
      redacted.Set(field, "[REDACTED]");
    }
  }
  
  return redacted;
}

std::string Logger::GenerateCorrelationId() {
  return base::StringPrintf("%08x-%04x-%04x-%04x-%012x",
                            base::RandUint64() & 0xffffffff,
                            base::RandUint64() & 0xffff,
                            base::RandUint64() & 0xffff,
                            base::RandUint64() & 0xffff,
                            base::RandUint64() & 0xffffffffffff);
}

}  // namespace diagnostics
}  // namespace toubkal
